### PR TITLE
New feature: synergy

### LINKED ISSUE #123: New feature: race of hero
> Prerequisite. This feature requires that [More gas efficient](https://github.com/HelheimLabs/autochessia/issues/84) is resolved.

### racial effect

What we want to have is that when a player has a certain number of heroes of the same race, it will give a specific effect to all his heroes. And this mechanism can give birth to a variety of hero combinations.

### race representation

```
Creature: {
  keySchema: {
    index: "uint16",
  },
  schema: {
    race: "uint32",
    health: "uint32",
    attack: "uint32",
    range: "uint32",
    defense: "uint32",
    speed: "uint32",
    movement: "uint32",
  },
},
```

`race` is an uint32 and is composited by 8 numbers of length 4 bits. Each number of length 4 bits represents one race and its position within `race` is fixed. 

| race     | race_1  | race_2-7 | race_8 |
|--------|--------|----------|--------|
| length  | 4 bits   | 24 bits    | 4 bits   |

`race_*` = `0000` or `0001`. `0001` represents this creature belongs to `race_*`

### PR DESCRIPTION
**Pull request by ClaudeZsb**
### Race synergy

| Race Synergy | number | description | effect index |
|---------------|---------|------------|-------------|
| Troll | 2 | 10% attack twice | 0x1c80 |
| Pandaren | 2 | +20% evasion | 0x8083 |
| | 4 | +30% evasion | 0x8084 |
| Orc | 2 | +100 max HP | 0x8080 |
| | 4 | +300 max HP | 0x8081 |
| Human | 2 | +15% damage reduction if at least one ally around | 0x0c80 |
| | 4 | +30% damage reduction if at least one ally around | 0x0c81 |
| God | 2 | +20% attack | 0x8082 |

### Class synergy

| Class Synergy | number | description | effect index |
|---------------|---------|------------|-------------|
| Knight | 2 | +10% immunity | 0x8104 |
| Warlock | 2 | 10% stun the enemy attacked | 0x1d00 |
| Assassin | 2 | +10% crit(x2) | 0x8105 |
| | 4 | +20% crit(x2) | 0x8106 |
| Warrior | 2 | +5 defense | 0x8102 |
| | 4 | +10 defense | 0x8103 |
| Mage | 2 | -20% enemy defense | 0x8100 |
| | 4 | -40% enemy defense | 0x8101 |

### Effect key

| Effect | with attribute modifier | registered event type | registered role in event | place holder | index |
|-------|-----------------------|-----------------------|------------------------|--------------|------|
| 0x8080 | 1 | 0000 | 0 | 001000 | 0 |
| 0x8081 | 1 | 0000 | 0 | 001000 | 1 |

### CODE REVIEW

File: packages/contracts/.prettierrc [hunk 1]
```diff
@@ -2,7 +2,7 @@
   "plugins": ["prettier-plugin-solidity"],
   "printWidth": 120,
   "semi": true,
-  "tabWidth": 2,
```
Comment (noyyyy) [inline packages/contracts/.prettierrc]: you can just set use forge to format solidity code, rather change prettier config

**Comment by noyyyy** [comment]
There are many break changes, I think we can merge it and take a look back on it if there is any problem. @aLIEzsss4

File: packages/game-balance/synergy.csv [hunk 2]
```diff
@@ -3,7 +3,7 @@ Orc,Species,Increase all max HP,100-300,2-4
 Troll,Species,All have a rate of x% to attack twice,10,2
 Pandaren,Species,Increase all rate of evasion by X%,20-30,2-4
 Human,Species,All descrease x% damage if it’s not alone(another piece in its nine-grid box),,
```
Comment (noyyyy) [inline packages/game-balance/synergy.csv]: I think you can add the number here and #163 can be closed.

Comment (ClaudeZsb) [reply packages/game-balance/synergy.csv]: ok

File: packages/contracts/mud.config.ts [hunk 3]
```diff
@@ -1,199 +1,248 @@
 import { mudConfig } from "@latticexyz/world/register";
 
 export default mudConfig({
```
Comment (aLIEzsss4) [inline packages/contracts/mud.config.ts:3]: Could you please format the mud.config.ts file in its original way, as it's hard to see the differences now.

Comment (ClaudeZsb) [reply packages/contracts/mud.config.ts:3]: done

**Comment by noyyyy** [comment]
wen merge

**Comment by aLIEzsss4** [comment]
> wen merge

now

**Review by aLIEzsss4 — APPROVED** [review]

**Pull request merged**